File: labs/lab2.3/csharp/TextLine.cs
// Клас-рядок, який містить методи, необхідні для роботи
// методів класу-контейнера.
using System.Text;

namespace lab2_3;

class TextLine : IWhiteSpaceRemover {
    private const string CONSONANTS = "qwrtypsdfghjklzxcvbnm";
    private const string PUNCTUATION = ".,!?&:;";

    public string Value {get; private set;}
    public int Length {get => this.Value.Length;}
    public int Consonants {get;} = 0;
    public char this[int key] {get => this.Value[key];}

    public TextLine() => this.Value = "";
    public TextLine(TextLine textLine) {
        this.Value = textLine.Value;
        this.Consonants = textLine.Consonants;
    }
    public TextLine(string text) {
        this.Value = text;
        foreach (char i in text) {
            if (CONSONANTS.Contains(char.ToLower(i))) this.Consonants++;
        }
    }
    
    public float ConsonantRatio() => (float)this.Consonants / this.Length;
    public void RemovePrefixWhiteSpace() {
        int i;
        for (i = 0; i < this.Value.Length; i++) {
            if (this.Value[i] != ' ') break;
        }
        this.Value = this.Value.Substring(i);
    }
    public void RemovePostfixWhiteSpace() {
        int i;
        for (i = this.Value.Length - 1; i > -1; i--) {
            if (this.Value[i] != ' ') break;
        }
        this.Value = this.Value.Substring(0,i);
    }
    public void RemoveExtraSpaces() {
        StringBuilder result = new();
        bool isWord = false;
        for (int i = 0; i < this.Value.Length; i++) {
            char currectChar = this.Value[i];
            if (currectChar == ' ') {
                char nextChar = '\0';
                for (int j = i + 1; j < this.Value.Length; j++) {
                    if (this.Value[j] != ' ') {
                        nextChar = this.Value[j];
                        i = j - 1;
                        break;    
                    }
                }
                if (isWord && nextChar != '\0' && !PUNCTUATION.Contains(nextChar)) {
                    result.Append(' ');
                    isWord = false;
                }
            } else {
                result.Append(currectChar);
                isWord = true;
            }
        }
        this.Value = result.ToString();
    }

    public override string ToString() => this.Value;
}

File: labs/lab2.3/cpp/TextLine.cpp
#include "headers/TextLine.h"
#include "algorithm" // for std::find
#include "sstream" // for std::ostringstream

bool contains(const string &arr, char element) {
    return std::find(arr.begin(),arr.end(),element) != arr.end();
}

TextLine::TextLine() : value("") {}
TextLine::TextLine(const string &text) : value(text) {
    for (char i : text) {
        if (contains(CONSONANTS,std::tolower(i))) {
            this->consonants++;
        }
    }
}
TextLine::TextLine(const TextLine &other) : value(other.value), consonants(other.consonants) {}
float TextLine::ConsonantRatio() const {
    return (float)this->consonants / this->value.length();
}
string TextLine::ToString() const {
    return this->value;
}
void TextLine::RemovePrefixWhiteSpace() {
    int i;
    for (i = 0; i < this->value.length(); i++) {
        if (this->value[i] != ' ') break;
    }
    this->value = this->value.substr(i);
}
void TextLine::RemovePostfixWhiteSpace() {
    int i;
    for (i = this->value.length() - 1; i > -1; i--) {
        if (this->value[i] != ' ') break;
    }
    this->value = this->value.substr(0,i);
}
void TextLine::RemoveExtraSpaces() {
    std::ostringstream result;
    bool isWord = false;
    for (int i = 0; i < this->value.length(); i++) {
        char currentChar = this->value[i];
        if (currentChar == ' ') {
            char nextCharacter = '\0';
            for (int j = i + 1; j < this->value.length(); j++) {
                if (this->value[j] != ' ') {
                    nextCharacter = this->value[j];
                    i = j - 1;
                    break;
                }
            }
            if (isWord && nextCharacter != '\0' && !contains(PUNCTUATION,nextCharacter)) {
                result << ' ';
                isWord = false;
            }
        } else {
            result << currentChar;
            isWord = true;
        }
    }
    this->value = result.str();
}
char TextLine::operator[](int key) const {
    return this->value[key];
}
TextLine& TextLine::operator=(const TextLine &other) {
    return *this;
}

//getters
int TextLine::GetConsonantCount() const { return this->consonants; }
int TextLine::Length() const { return this->value.length(); }